File: schemas/analytics-metrics/convex/experiment_variants.ts
// experiment_variants: Variants within experiments.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const experiment_variants = defineTable({
  experimentId: v.id("experiments"),
  name: v.string(),
  description: v.optional(v.string()),
  isControl: v.boolean(),
  weight: v.number(),
  config: v.optional(v.any()),
})
  .index("by_experiment_id_and_name", ["experimentId", "name"]);

File: schemas/analytics-metrics/spacetimedb/experiment_variants.rs
// experiment_variants: Variant arms within an experiment with weight allocation.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite unique constraint (not expressible inline):
// - unique(experiment_id, name)

#[spacetimedb::table(name = experiment_variants, public)]
pub struct ExperimentVariant {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub experiment_id: String, // UUID, FK → experiments.id (cascade delete)

    pub name: String,
    pub description: Option<String>,
    pub is_control: bool,
    pub weight: f64,
    pub config: Option<String>, // JSON
    pub created_at: Timestamp,
}
